Update make_deb.py to work for Python2 or 3

aiuto opened the pull request:
Generally being more careful about unicode handling.
Some very weird magic with pre-processing argv to undo undesired interpreter behavior.

This leaves the code in a state where someone could use either interpreter.
Next step it to figure out how to select the python version to match whatever toolchain we have.

aiuto commented:
Mostly fixes #35 
@qzmfranklin, @jwnimmer-tri  Do you want to patch this in, change the target to PY3 and see if it works for you?

brandjon commented:
> Next step it to figure out how to select the python version to match whatever toolchain we have.

#8713. I think the easiest thing is to make it `python_version = "PY3"` since that's preferred, but keep it compatible with PY2 (enforced via a `python_version = "PY2"` `py_test`). In the host config you don't control its real Python version anyway.

If a user doesn't have Python 3 installed but wants to run targets marked PY3 that are in reality compatible with both Python 2 and 3, they should use the non-strict toolchain (implemented in #8547).

brandjon commented on pkg/make_deb.py:
```diff
@@ -303,15 +303,30 @@ def CreateChanges(output,
     changes_fh.write(changesdata.encode('utf-8'))
 
 
 def GetFlagValue(flagvalue, strip=True):
```
Nit: The name's a bit vague. Maybe rename to NormalizeUnicode or NormalizeString or something? The docstring should clarify what specific kind of normalization we're talking about.

aiuto replied:
I want to keep the name of GetFlagValue just to limit the number of changed lines.
But I did expand the docstring and substantially re-wrote it to do the fsencoding stuff here, so that we do not (incorrectly) turn a utf-8 path in @filename args to an improper path.

aiuto commented:
Also changed the default version to py3 and added a py2 test.

brandjon commented on pkg/BUILD:
```diff
@@ -42,11 +42,22 @@ py_binary(
     name = "make_deb",
     srcs = ["make_deb.py"],
+    python_version = "PY3",
+    srcs_version = "PY3",
```
I'd actually leave out `srcs_version` entirely and let it default to `PY2AND3`. The source code is compatible with both versions, after all. It's only the target that says to use one or the other.

If this were a `py_library` target, `PY2AND3` would be necessary in order for the target to be usable as a dependency in binaries of both versions. It doesn't matter for `py_binary` except for documentation / clarity.

Actually, setting it to the default `PY2AND3` value explicitly could even be helpful, in that it replaces the "This works for both PY2 and PY3" comment. Same principle for the `make_deb_py2` target.

aiuto replied:
Done.

brandjon commented on pkg/make_deb.py:
```diff
@@ -386,4 +400,5 @@ def main(unused_argv):
   # https://docs.python.org/3/library/os.html#os.fsencode
   # https://www.python.org/dev/peps/pep-0383/
-  main(FLAGS([os.fsencode(arg).decode('utf-8') for arg in sys.argv]))
+  # main(FLAGS([os.fsencode(arg).decode('utf-8') for arg in sys.argv]))
```
Stale line. Also should the above comment be deleted now that the logic's moved to `GetFlagValue`?

brandjon commented on pkg/make_deb.py:
```diff
@@ -308,26 +308,40 @@ def CreateChanges(output,
         flagvalue = f.read().decode('utf-8')
+    else:
+      # convert fs specific encoding back to proper unicode.
+      flagvalue = os.fsencode(flagvalue).decode('utf-8')
```
`fsencode()` will return the original bytes. If we're on Windows, what encoding will those bytes be? If they're not utf-8, doesn't that mean we'll be mis-decoding `flagvalue`?

Another way to put it: `sys.argv` gives us a sequence of unicode code points as specified by the system. On windows this can actually be interesting characters (>255), while on linux this is always characters in range 0-255. This is because the use of utf-8 in linux is a convention, not a built-in feature understood by the OS. Therefore, unconditionally decoding the bytes as utf-8 is (I think?) incorrect on systems that follow a different convention.

aiuto commented:
> Mostly fixes #35
> @qzmfranklin, @jwnimmer-tri Do you want to patch this in, change the target to PY3 and see if it works for you?

I added changed to PY3 as the default and added a py2 test, so this is less important. I think I will merge and we'll see how that goes.

jwnimmer-tri commented:
Thanks for your work on this!  Apologies for the slow reply, I have been on vacation.

I tested the latest master here 96e34abe84353387fd378fba48be88836f6c3f2d against my tree, and I can confirm that the `decode()`-related problems I saw in bazelbuild/bazel/issues/7370 are resolved.
